vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+  m_socket.Closed({ this, &WinRTWebSocketResource::OnClosed });
+
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::Untrusted);
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::InvalidName);
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
+}
+
+#pragma region Private members
+
+fire_and_forget WinRTWebSocketResource::PerformConnect()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    co_await m_socket.ConnectAsync(m_uri);
+
+    if (m_connectHandler)
+    {
+      m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    hresult_error e{ hr };
+    //OutputDebugString("WebSocket.connect failed (0x%8X) %ls\n", e);
+
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  // m_connectPerformed.Set();
+  m_connectPerformed.set_value();
+  m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    m_writer.WriteBytes(arr);
+
+    co_await m_writer.StoreAsync();
+
+    if (m_pingHandler)
+    {
+      m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(hresult_error(hr).message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  if (m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    string message;
+    bool isBinary;
+    size_t length;
+    std::tie(message, isBinary) = m_writeQueue.front();
+    m_writeQueue.pop();
+
+    if (isBinary)
+    {
+      m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      length = message.size();
```

Comment:
>size [](start = 23, length = 4)

(nit) please use either "size" or "length" consistently

Reply:
Can't.

When dealing with a binary buffer, the function called is `Length()`. When dealing with the UTF-8 array, it's `size()`.

Ultimately, this variable is passed to the `m_writeHandler` callback.

Reply:
I meant "message.size()"
In line 173, it's "message.size(), in line 176, it's "message.length()" you should be able to just stick to single API, right?

---
In reply to: [394018513](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r394018513) [](ancestors = 394018513)